metrics/ci_integrations/docs/buildkite/diagrams/buildkite_client_class_diagram.puml

```diff
@@ -0,0 +1,101 @@
+
+package integration.interface.base.client.mapper {
+    interface Mapper<SourceType, DestinationType> {
+        map(SourceType value): DestinationType
+        unmap(DestinationType value): SourceType
+    }
+}
+
+package buildkite {
+    package models {
+        class BuildkiteBuild {
+            + id: int
+            + number: int
+            + state: BuildkiteBuildState
+            + webUrl: String
+            + startedAt: DateTime
+            + finishedAt: DateTime
+
+            + factory fromJson(Map<String, dynamic> json): BuildkiteBuild
+            + toJson(): Map<String, dynamic>
+            + listFromJson(List<dynamic> list): List<BuildkiteBuild>
+        }
+
+        class BuildkiteArtifact {
+            + id: int
+            + filename: String
+            + downloadUrl: String
+            + mimeType: String
+
+            + factory fromJson(Map<String, dynamic> json): BuildkiteArtifact
+            + toJson(): Map<String, dynamic>
+            + listFromJson(List<dynamic> list): List<BuildkiteArtifact>
+        }
+
+        interface Page<T> {
+            + page: int
+            + perPage: int
+            + nextPageUrl: String
+            + values: List<T>
+        }
+
+        class BuildkiteBuildsPage<BuildkiteBuild> implements Page {
+            + page: int
+            + perPage: int
+            + nextPageUrl: String
+            + values: List<BuildkiteBuild>
+        }
+
+        class BuildkiteArtifactsPage<BuildkiteArtifact> implements Page {
+            + page: int
+            + perPage: int
+            + nextPageUrl: String
+            + values: List<BuildkiteArtifact>
+        }
+
+        enum BuildkiteBuildState {
+          running,
+          scheduled,
+          passed,
+          failed,
+          blocked,
+          canceled,
+          canceling,
+          skipped,
+          notRun,
+          finished
+        }
+    }
+
+    package mappers {
+        class BuildkiteBuildStateMapper<String, BuildkiteBuildState> implements Mapper {
+            + map(String state): BuildkiteBuildState
+            + unmap(BuildkiteBuildState state): String
+        }
+
+    }
+
+    class BuildkiteClient {
+        + buildkiteApiUrl: String
+        + organizationSlug: String
+        + authorization: AuthorizationBase
+
+        + getHeaders(): Map<String, String>
+        + fetchBuildkiteBuilds(String pipelineSlug, BuildkiteBuildState state, DateTime finishedFrom, int perPage, int page): Future<InteractionResult<BuildkiteBuildsPage>>
+        + fetchBuildkiteBuildsNext(BuildkiteBuildsPage currentPage): Future<InteractionResult<BuildkiteBuildsPage>>
+        + fetchBuildkiteArtifacts(int buildNumber, int perPage, int page): Future<InteractionResult<BuildkiteArtifactsPage>
+        + fetchBuildkiteArtifactsNext(BuildkiteArtifactsPage currentPage): Future<InteractionResult<BuildkiteArtifactsPage>
+        + downloadBuildkiteArtifact(String url): Future<InteractionResult<UInt8List>>
```

Comment:
As a point to discuss, maybe it would be better to avoid using `Future<InteractionResult<...>>` in diagrams as this seems to be not necessary for understanding the design and architecture. It makes the diagram more complex and hard-to-read but doesn't really useful for the main idea and the purpose of this diagram. 

cc @illia-romanenko @solid-vovabeloded

Reply:
As we discussed let's skip this suggestion for now and discuss it later on Friday. I mark this as resolved for now.